linera-execution/src/lib.rs

```diff
@@ -50,6 +51,72 @@ use serde::{Deserialize, Serialize};
 use std::{io, path::Path, str::FromStr, sync::Arc};
 use thiserror::Error;
 
+/// substract an amount to a balence and report an error if that is impossible
+pub fn sub_assign_fees(balance: &mut Amount, fees: Amount) -> Result<(), PricingError> {
+    Ok(balance
+        .try_sub_assign(fees)
+        .map_err(|_| ArithmeticError::Underflow)?)
+}
+
+/// The entries of the runtime related to fuel and storage
+#[derive(Copy, Debug, Clone)]
+pub struct RuntimeLimits {
+    /// The maximum size of read allowed per block
+    pub maximum_bytes_read: u64,
+    /// The maximum size of write allowed per block
+    pub maximum_bytes_written: u64,
+}
+
+impl Default for RuntimeLimits {
+    fn default() -> Self {
+        RuntimeLimits {
+            maximum_bytes_read: u64::MAX,
+            maximum_bytes_written: u64::MAX,
+        }
+    }
+}
+
+/// The entries of the runtime related to fuel and storage
+#[derive(Copy, Debug, Clone)]
+pub struct RuntimeCounts {
+    /// The remaining fuel available
+    pub remaining_fuel: u64,
+    /// The number of read operations
+    pub num_reads: u64,
+    /// The bytes that have been read
+    pub bytes_read: u64,
+    /// The bytes that have been written
+    pub bytes_written: u64,
+}
+
+pub fn update_limits(
+    balance: &mut Amount,
+    runtime_limits: &mut RuntimeLimits,
+    pricing: &Pricing,
+    runtime_counts: RuntimeCounts,
+) -> Result<(), ExecutionError> {
+    let initial_fuel = pricing.remaining_fuel(balance.clone());
+    let used_fuel = initial_fuel.saturating_sub(runtime_counts.remaining_fuel);
+    sub_assign_fees(balance, pricing.fuel_price(used_fuel)?)?;
+    sub_assign_fees(
+        balance,
+        pricing.storage_num_reads_price(&runtime_counts.num_reads)?,
+    )?;
+    let bytes_read = runtime_counts.bytes_read;
+    runtime_limits.maximum_bytes_read -= bytes_read;
```

Comment:
We should either use `checked_sub` everywhere, or add documentation to this function that the caller needs to ensure that the counts don't exceed the limits.

Reply:
I added the documentation.